### ISSUE TITLE
Error performing request: dial tcp: i/o timeout

### ISSUE THREAD

**Issue by olzorenko** [cloudfoundry/cf-release#923]
We've faced an issue when running smoke/acceptance tests on CF 231 version:

Here is an chunk of logs (domain and password was blinked):

```
Running smoke tests...
/var/vcap/data/packages/golang1.4/714698bc352d2a1dbe321376f0676037568147bb.1-f27d8d9c7173982dabcc80d3f1b8c5244f4d4b89/bin/go
[1457081026] CF-Logging-Smoke-Tests - 2/2 specs 
[2016-03-04 08:43:49.80 (UTC)]> cf api api.cf.DOMAIN_NAME --skip-ssl-validation 
Setting api endpoint to api.cf.DOMAIN_NAME...
OK

                   
API endpoint:   https://api.cf.DOMAIN_NAME (API version: 2.51.0)   
Not logged in. Use 'cf login' to log in.

[2016-03-04 08:43:50.00 (UTC)]> cf auth admin PASS 
API endpoint: https://api.cf.DOMAIN_NAME
Authenticating...
OK
Use 'cf target' to view or set your target org and space

[2016-03-04 08:43:55.46 (UTC)]> cf create-quota cf-smoke-test_QUOTA -m 10G -r 10 -s 2 
Creating quota cf-smoke-test_QUOTA as admin...
OK

[2016-03-04 08:43:55.58 (UTC)]> cf create-org cf-smoke-test 
Creating org cf-smoke-test as admin...
OK

Assigning role OrgManager to user admin in org cf-smoke-test ...
OK

TIP: Use 'cf target -o cf-smoke-test' to target new org

[2016-03-04 08:44:10.94 (UTC)]> cf set-quota cf-smoke-test cf-smoke-test_QUOTA 
Setting quota cf-smoke-test_QUOTA to org cf-smoke-test as admin...
OK

[2016-03-04 08:44:11.13 (UTC)]> cf target -o cf-smoke-test 
                   
API endpoint:   https://api.cf.DOMAIN_NAME (API version: 2.51.0)   
User:           admin   
Org:            cf-smoke-test   
Space:          No space targeted, use 'cf target -s SPACE'   

[2016-03-04 08:44:16.29 (UTC)]> cf create-space -o cf-smoke-test cf-smoke-test 
Creating space cf-smoke-test in org cf-smoke-test as admin...
FAILED
Error finding org cf-smoke-test
Error performing request: Get https://api.cf.DOMAIN_NAME/v2/organizations?q=name%!A(MISSING)cf-smoke-test&inline-relations-depth=1: dial tcp: i/o timeout

[2016-03-04 08:44:31.40 (UTC)]> cf delete-space cf-smoke-test -f 
FAILED
Space cf-smoke-test not found

------------------------------
• Failure in Spec Setup (BeforeEach) [46.720 seconds]
[BeforeEach] Loggregator: cf logs linux can see app messages in the logs 
/var/vcap/packages/smoke-tests/src/github.com/cloudfoundry/cf-smoke-tests/smoke/logging/loggregator_test.go:40

  Expected
      <int>: 1
  to match exit code:
      <int>: 0
```

We're running CF on vSphere. Also this problem appears when we're trying running this CF version locally via bosh-lite and also when running smoke/acceptance tests.

Any help and thoughts will be greatly appreciated!

Best Regards,
Oleksii.

**Comment by ndemeshchenko** [comment]
Same here, issue is happening randomly and not clear on what have caused it.
It can fail on accessing CF Api endpoint on one request and run success on next. Can u guys, please advise where can I check for any deep investigation? thanks

**Comment by vital-st1x** [comment]
The same issue for new 231 version (webdav blobstore or nfs). Clean instructions for debug (not only CF_TRACE=true) will be nice.

**Comment by robdimsdale** [comment]
We have recently started to observe this too on bosh-lite running on AWS - the CLI is hard-coded to time out after 5 seconds, so it is possible that it is taking more than 5 seconds for some requests to reach the API.

However, that code hasn't changed in 18 months (and it isn't confined to a single infrastructure so we can't assume it's a problem only at the IAAS), so we are investigating whether some other component has started to slow down.

We will update this issue with our findings.

Thanks,
Rob && @dsabeti

**Comment by robdimsdale** [comment]
@vital-st1x @ndemeshchenko @olzorenko our issue was a misconfigured DNS record in `/etc/resolv.conf` so we suggest you check your nameservers point to valid DNS servers, and that those servers have the required DNS entries for the hostnames.

Also potentially related is that the CLI recently switched from golang 1.5 to 1.6 - see https://github.com/cloudfoundry/cli/issues/763.

Let us know if fixing your DNS and nameservers solves the issues you are having.

Thanks,
Rob && @APShirley

**Comment by olzorenko** [comment]
Hello robdimsdale!

Thank you for your reply!

I'll check your suggestion about DNS configuration and get back to you shortly.

Btw, here are my bosh and cf versions:

BOSH 1.3202.0
cf version 6.15.0+fa1bfe2-2016-01-13

Best Regards,
Oleksii.

**Comment by olzorenko** [comment]
Hello @robdimsdale!

Sorry, but I can't find a place where I can configure Cloud Foundry's DNS service.

Do I need to do this configuration on my global DNS (which serves my domain) or there are a separate job in bosh or component in cloud foundry deployment?

Simply put - where I need to check/modify /etc/resolv.conf for correct nameservers?

Thanks in advance!

Best Regards,
Oleksii.

**Comment by robdimsdale** [comment]
@olzorenko each VM has its own `/etc/resolv.conf` which is written by bosh when the VM is created. The IAAS can also add entries, e.g. AWS DHCP options. If a job is co-located with the `consul_agent` job, the `/etc/resolv.conf` may be modified after the VM is created.

In case you're not familiar; if you run an errand with `--keep-alive` then bosh will keep the VM around after the errand has exited, giving you the opportunity to look around the VM afterwards.

Once you get onto the VM, could you please show us the contents of `/etc/resolv.conf`?

For example, on a local bosh-lite, what we see is:

```
bosh_wxakgf1cm@b6dc01c5-eb42-4231-8870-1043b5ababed:~$ cat /etc/resolv.conf
nameserver 127.0.0.1
nameserver 8.8.8.8
nameserver 10.0.2.3
```

As you can see, the first nameserver that is used will be `127.0.0.1`. On a bosh-lite VM this will ultimately use host machine's config, but this will vary across IAASs. If using this config takes too long to resolve a domain name, the CF CLI will timeout. As a temporary workaround, you could reorder the entries in `/etc/resolv.conf` and re-rerun the errand; a nameserver that responds with the DNS entry faster than 5 seconds will not cause the CLI to fail (or at least, it will fail for a different reason).

The underlying issue is probably that one of your nameserver entries points to either:
- an IP that isn't actually a DNS server
- an IP that takes too long to get a round-trip response to the DNS query.

We found ourselves in the former situation due to a misconfigured IAAS.

Rob & @nimakaviani

**Comment by jona23** [comment]
@robdimsdale Here I am Rob, I'm Giovanni, the user that posted the issue about "dial: tcp i/o timeout" error on lists.cloudfoundry.org. I have many news about my problems.
I'm working on a BOSH-Lite deployment of Cloud Foundry with a group of research engineers in Italy, specifically my team is a research engineer group of an italian university. The thing is that the university has blocked all DNS traffic outside of a specif DNS, that is the DNS used for the entire universty (its IP is 193.205.160.3). So, every DNS query done by the VMs to 8.8.8.8 has no response, infact i sniffed the traffic with Wireshark and 8.8.8.8:53 doesn't respond. Moreover, almost in all of 13 VMs deploy by bosh there the same nameserver entries you posted:
`nameserver 127.0.0.1
nameserver 8.8.8.8
nameserver 10.0.2.3`
or
`nameserver 8.8.8.8
nameserver 10.0.2.3`

I'm saying thath because even if I change or delete the entry with the public DNS IP 8.8.8.8, the issue remains. Also, sniffing the traffic after the change, there are DNS queries to 8.8.8.8 yet. That's really a strange thing.

However, i think that the problem is about the hard-coded implementation of `gorouter.go`. Infatct, last day, using my cell phone hotspot to push a simple app to the cloud, I had the same "dial: tcp i/o timeout" error that I already had with the wired network. Today, indeed, the problem disappeared (always using the hotspot connecction).
This behaviour made me think that the only reason for having this issue is the hard-coded timeout. What do you think about changing the `gorouter.go` file, to fix the error? Is there possibile that compiling a modded version of cli, the issue could be solved?

**Comment by APShirley** [comment]
@jona23,
Can you do a `dig` on `api.bosh-lite.com` from any vm in your deployment and post the output here. 

`10.0.2.3` is the `bosh-director` gateway that should eventually resolve to your university dns server. If you have a dns server that takes too long to respond, or is incorrect the timeout issue will manifest itself. 

Our hypothesis in this situation is that, if you have removed the public dns `8.8.8.8` and you still face the timeout problem, it could be that your university dns is taking too long to respond. However, the fact that you still see traffic to `8.8.8.8` after removing the ip from `/etc/resolv.conf` makes us think that either the dns settings from the vagrant box are different than those of the host machine which might still have a bad nameserver entry, or that there is some misconfiguration with your `bosh-director`. Can you check the `/etc/resolv.conf` of the director, and the host machine for entries that can't be reached from inside the university network? If `8.8.8.8` appears first, it is possible that deleting or moving the entry down will fix the issue.

We have reached out to the bosh team to investigate the possibility of bad dns settings being an issue between the vagrant box and the host machine. We will update you on that. In the meantime, the results of `dig api.bosh-lite.com` would probably be helpful in figuring out what is going on with the routing. 

Alternatively, if you are willing to modify the behavior of your gorouter.go and compile a modified version of the cf-cli this might fix the problem you are seeing. The cf-cli team are aware of the issue, and there is a story in their backlog to make the timeout configurable but it hasn't been prioritized yet. 

Thanks,

Al, @robdimsdale, & @nimakaviani

**Comment by gnapoli2** [comment]
@APShirley ,
here's the output of `dig api.bosh-lite.com`:

```
bosh_j627v2heg@c7656e4b-f5ba-43b0-8599-bd5e0f7dcde4:~$ dig api.bosh-lite.com
;; Warning: Message parser reports malformed message packet.

; <<>> DiG 9.9.5-3ubuntu0.5-Ubuntu <<>> api.bosh-lite.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 39660
;; flags: qr aa rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; QUESTION SECTION:
;api.bosh-lite.com.		IN	A

;; ANSWER SECTION:
api.bosh-lite.com.	3600	IN	A	10.244.0.34

;; Query time: 152 msec
;; SERVER: 10.0.2.3#53(10.0.2.3)
;; WHEN: Thu Mar 17 09:11:46 UTC 2016
;; MSG SIZE  rcvd: 51

```
I have launched it on the VM `api_z1/0`.

About the `resolv.conf` files of the `bosh-director` and host, here what's inside them:

- `bosh-director`
```
nameserver 8.8.8.8
nameserver 10.0.2.3

```

- about my `host`, i have to say that i'm using the `network-manger` service, so maybe `resolv.conf`file is ignored by it. However, university DNS ip server is properly set to `193.205.160.3` in the `network-manager` configuration, and the `resolv.conf` file only contains
```
nameserver 127.0.0.1

```
Thinking about the configuration of `/etc/resolv.conf` in the `bosh-director`, maybe that is the problem about the DNS traffic to the `8.8.8.8` public ip address, even if i remove the `8.8.8.8` entry in all the VMs deployed. So, i'll try to force the vagrant box using the my host resolver through
```
VBoxManage modifyvm "VM name" --natdnshostresolver1 on

```
or changing the Vagrantfile of `bosh-lite`.

Moreover, digging in the web i saw that many users have this particular issue. Is there any way to ask to `cf-cli` mantaining team to fix the `gateway.go` file by changing the timeout time? I think that severel problems will be solved.

**Comment by APShirley** [comment]
Hi @gnapoli2,
Please do try removing the 8.8.8.8 nameserver from the bosh-director. Also, as a temporary workaround, as already suggested in [this issue thread](https://github.com/cloudfoundry/cli/issues/763), you can try setting the global environment variable `GODEBUG=netdns=cgo` to see if that helps (it is a temporary workaround suggested by the cf-cli team). Also when running `cf push` please set `CF_TRACE=true` to help us get a trace for debugging the issue.

We are going to open a [new issue](https://github.com/cloudfoundry/cli/issues/new) with the cf-cli team to highlight the problem. We will include the output of your `CF_TRACE` in the issue we intend to create. We also need more information on your setup to open the issue:

- Command (e.g. `cf push`)
- CLI Version (`6.15`, `6.16`)
- Error (Dial tcp: i/o timeout, or any error you receive)
- Stack Trace
- CC API Version
- CF Trace 
- Platform Details (e.g. what version of Ubuntu?, 64-bit?)
- Shell (sh, bash, zsh...)

We (@robdimsdale) created an [un-prioritized story](https://www.pivotaltracker.com/story/show/115252237) to work on it in their tracker. Which we can point to, to highlight that this is a problem for many users.

Thanks for all your input,

Al & @nimakaviani

**Comment by gnapoli2** [comment]
Sorry, i didn't see the message. Tomorrow (here in Italy Italy 19:00 pm) i'll post all the informazioni you asked me.
However i did another try: i changed the gateway of my connection to one that i'm using alone and seems that the issue  disappeared. So my answer is;
- do you think that the gateway was the problem?(it's used for all the località network in my laboratory so responding to many requests it brought me this issue)
- or do you think that a sort of DNS località cache has been saved and then the requests are done?

Thank you for your support.

**Comment by gnapoli2** [comment]
Hi @APShirley , here's the informations:

- Command: `cf push`, `cf apps`, `cf domains`, `cf orgs`, and other `cf` query (in poor words every http request made with `cf-cli`;

- CLI version: `cf version 6.16.1+924508c-2016-02-26`;

- Error: `Error performing GET ... dial tcp: i/o timeout` or `Error performing POST ... dial tcp: i/o timeout`;

- CC Api Version: `v2.52.0`;

- CF Trace: about the CF_Trace=true, is useless to post the result because the error comes with different kind of request as i said (GET/POST);

- Platform Details: `Ubuntu 64 bit 14.04 LTS`;

- Shell: `bash`

**Comment by APShirley** [comment]
Hi @gnapoli2,

I have created a [new issue](https://github.com/cloudfoundry/cli/issues/802) with the cf api team. 
As for your questions around what we think this is

- "do you think that the gateway was the problem?"

Quite possibly, it could be the gateway took too long to respond (I doubt this though), we feel it is more likely that the gateway is in some way misconfigured (or configured in conflict with other dns settings) and is trying to reach `8.8.8.8` which as we know, is blocked by your university sys admins.

- "or do you think that a sort of DNS local cache has been saved and then the requests are done?"

I don't know of any mechanism for DNS caching that would cause this but I'm not very familiar with the way linux kernels handle dns resolution, as far as I know `/etc/resolv.conf` should be the source of truth here. So as long as nowhere in the gateways used is an address trying to be resolved through the `8.8.8.8` dns server there shouldn't be an issue. Using `traceroute` might hold the answer here since you saw the address in the routing before.

We will update you as we know more about this issue, feel free to also check the issues on the cli team's github page, as there may be updates there. We will continue to investigate from our end. If you find any more information that might be relevant let us know. In the meantime I would suggest you look at the faulty gateway's configs, load averages, open connections etc... and possibly continue using the other gateway. 

Grazie,

Al

**Comment by robdimsdale** [comment]
@gnapoli2 @olzorenko we are going to close this issue now, and suggest you take a look at the issue we created on the CF CLI repo: https://github.com/cloudfoundry/cli/issues/802.

Thanks,
Rob & @APShirley
